### ISSUE TITLE
[Question/Feature?] Dynamically run global.setup authentications

### ISSUE THREAD

**Issue by bshostak** [microsoft/playwright#28257]
Hello, wanted to ask a quick question that i can't seem to find a solid answer to.

Currently, I have a automation project with tests that use many different users as auth states. These auth users differ because they are all from different countries so the tests I have require specific authStates to use. For example, a test focused on India flows will require the India User, and so on for each other country. My global setup looks something like this:

```typescript
import { mergeTests } from '@playwright/test';
import { test as ssoTest } from '@package/automation/sso-page-modules';
import { test as hubTest } from '@package/automation/hub-page-modules';
const test = mergeTests(ssoTest, hubTest);
import { testUsers } from '../Utils/TestUsers';

const TestConfig = require('./../TestConfig.json');

type AuthUsers = [userInfo: string, userId: string, authStatePath: string];

const userIds: AuthUsers[] = [
  ['baseuser', testUsers.baseuser, './AuthStates/baseuser.json'],
  ['baseusertwo', testUsers.baseusertwo, './AuthStates/baseusertwo.json'],
  ['caUser', testUsers.caUser, './AuthStates/caUser.json'],
  ['mxUser', testUsers.mxUser, './AuthStates/mxUser.json'],
  ['auUser', testUsers.auUser, './AuthStates/auUser.json'],
  ['nzUser', testUsers.nzUser, './AuthStates/nzUser.json'],
  ['coUser', testUsers.coUser, './AuthStates/coUser.json'],
  ['phUser', testUsers.phUser, './AuthStates/phUser.json'],
  ['twUser', testUsers.twUser, './AuthStates/twUser.json'],
  ['sgUser', testUsers.sgUser, './AuthStates/sgUser.json'],
  ['myUser', testUsers.myUser, './AuthStates/myUser.json'],
  ['krUser', testUsers.krUser, './AuthStates/krUser.json'],
  ['hkUser', testUsers.hkUser, './AuthStates/hkUser.json'],
  ['roUser', testUsers.roUser, './AuthStates/roUser.json'],
  ['inUser', testUsers.inUser, './AuthStates/inUser.json'],
];


test.describe('Global Setup', () => {
  userIds.forEach(async (data) => {
    test(`Setup Login for ${data[0]} ${data[1]}`, async ({ ssoLoginPage, hubHomePage, page }) => {
      await ssoLoginPage.get(TestConfig.targetEnv);
      await ssoLoginPage.login(data[1], await getPassword(data[1]));
      await page.context().storageState({ path: data[2] });
    });
  });
});
```

So before each test run I have to run through all the auth states. If I run a test file that only requires the authState for the India user, I still have to wait for all the auth states to complete. If I run a test file that requires NO auth states, it will still run through all the auth states because of how global.setup.ts works. 

Config for reference:

```typescript
const config: PlaywrightTestConfig = {
  testDir: './tests',
  /* Maximum time one test can run for. */
  timeout: 30 * 1000,
  expect: {
    /**
     * Maximum time expect() should wait for the condition to be met.
     * For example in `await expect(locator).toHaveText();`
     */
    timeout: 10000,
  },
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only. Due to environment flakiness, there may be times when we need to do a retry, so I have set max to 2 */
  retries: process.env.CI ? 2 : 1,

  workers: 4,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: [
    ['html', { open: 'never' }],
  ],
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Maximum time each action such as `click()` can take. Defaults to 0 (no limit). */
    actionTimeout: 0,
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://localhost:3000',

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'retain-on-failure',
    ignoreHTTPSErrors: true,
    video: 'retain-on-failure',
  },

  /* 
  Configure projects for major browsers. We are setting our viewport size to be more of a desktop size view.
  Future specific Mobile size views will need to be created as a new project. 
  */
  projects: [
    {
      name: 'setup',
      testMatch: /global.setup\.ts/,
    },
    {
      name: 'Chrome',
      use: {
        ...devices['Desktop Chrome'],
        viewport: { height: 1200, width: 1600 },
      },
      dependencies: ['setup'],
    },

    {
      name: 'Firefox',
      use: {
        ...devices['Desktop Firefox'],
        viewport: { height: 1200, width: 1600 },
      },
      dependencies: ['setup'],
    },

    {
      name: 'Safari',
      use: {
        ...devices['Desktop Safari'],
        viewport: { height: 1200, width: 1600 },
      },
      dependencies: ['setup'],
    },



 
  ],

};
```

My question is: Is there a way to dynamically run specific auth states based on the tests that are being executed? So if I run:
```npx playwright test tests/inTests```
These tests all have:
```typescript
test.use({ storageState: './AuthStates/inUser.json' });
```
under the describe. But on run it will still go through all my other market authentications even though it's not needed for this test run. I would only need the one specified in the test.use(). If i am approaching this in a bad way, please point me to a better approach with using this many auth users.

Thank you for your time!

**Comment by yury-s** [comment]
You can split the setup tests into separate country-specific setups: ca-setup.ts, mx-setupt.ts, au-setup.ts etc and have corresponding country-specific tests depend only on the corresponding setup projects, something like:

```js
    {
      name: 'ca-setup',
      testMatch: /global.setup\.ts/,
    },
    {
      name: 'ca-chrome',
      use: {
        ...devices['Desktop Chrome'],
        viewport: { height: 1200, width: 1600 },
      },
      dependencies: ['ca-setup'],
    },
    {
      name: 'mx-setup',
      testMatch: /global.setup\.ts/,
    },
    {
      name: 'mx-chrome',
      use: {
        ...devices['Desktop Chrome'],
        viewport: { height: 1200, width: 1600 },
      },
      dependencies: ['ca-setup'],
    },
...
```
(you can extract common parts of the config into variables to reduce duplication, but the idea is the same). Would that work for you?

**Comment by bshostak** [comment]
Thanks @yury-s! I'm not sure I 100% follow the suggestion but I can definitely try to create a bigger project list corresponding to each country and then have a separate dependency for that specific country setup. I think I might also have to testMatch the country folder for the specific country projects if I'm not mistaken.

**Comment by yury-s** [comment]
> Thanks @yury-s! I'm not sure I 100% follow the suggestion but I can definitely try to create a bigger project list corresponding to each country and then have a separate dependency for that specific country setup. I think I might also have to testMatch the country folder for the specific country projects if I'm not mistaken.

Yes, you'll need to have separate testMatch filters for each country project. The idea is to have project(s) per country.

**Comment by bshostak** [comment]
@yury-s Thanks again for the suggestion. I refactored and implemented a function to build out the projects and it all works great and as i'd expect. Example:
```typescript
export function buildProject(countryCodes: Country[]) {
  return countryCodes.flatMap((countryCode) => {
    const viewPortSpecs = { height: 1200, width: 1600 };
    const testMatchDetails = new RegExp(`tests/${countryCode}/.*\\.spec\\.ts`);

    return [
      {
        name: `${countryCode}-setup`,
        testMatch: new RegExp(`globalSetups\/${countryCode}-global\\.setup\\.ts$`),
      },
      {
        name: `${countryCode}-chrome`,
        testMatch: testMatchDetails,
        use: {
          ...devices['Desktop Chrome'],
          viewport: viewPortSpecs,
        },
        dependencies: [`${countryCode}-setup`],
      },
      {
        name: `${countryCode}-firefox`,
        testMatch: testMatchDetails,
        use: {
          ...devices['Desktop Firefox'],
          viewport: viewPortSpecs,
        },
        dependencies: [`${countryCode}-setup`],
      },
      {
        name: `${countryCode}-safari`,
        testMatch: testMatchDetails,
      },
    ];
  });
}

export const projectCountries = [
  Country.AU,
  Country.CA,
  Country.CO,
  Country.IN,
  Country.MX,
  Country.NZ,
  Country.PH,
  Country.RO,
  Country.UK,
  Country.US,
];
```
